core/src/main/scala/org/apache/spark/sql/sources/SparkOptimizations.scala

```diff
@@ -0,0 +1,410 @@
+import org.apache.spark.sql.catalyst.planning.ExtractEquiJoinKeys
+import org.apache.spark.sql.catalyst.plans.logical.{LogicalPlan, SubqueryAlias}
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.execution.columnar.impl.{BaseColumnFormatRelation, ColumnFormatRelation, IndexColumnFormatRelation}
+import org.apache.spark.sql.execution.datasources.LogicalRelation
+
+
+/**
+  * Replace table with index hint
+  */
+case class ResolveQueryHints(snappySession: SnappySession) extends Rule[LogicalPlan]
+{
+  lazy val catalog = snappySession.sessionState.catalog
+
+  lazy val analyzer = snappySession.sessionState.analyzer
+
+  override def apply(plan: LogicalPlan): LogicalPlan = {
+
+    val explicitIndexHint = getIndexHints
+
+    if (explicitIndexHint.isEmpty) {
+      return plan
+    }
+
+    plan transformUp {
+        case table@LogicalRelation(colRelation: ColumnFormatRelation, _, _) =>
+          explicitIndexHint.get(colRelation.table).getOrElse(table)
+        case subQuery@SubqueryAlias(alias, LogicalRelation(_, _, _)) =>
+          explicitIndexHint.get(alias) match {
+            case Some(index) => SubqueryAlias(alias, index)
+            case _ => subQuery
+          }
+    } transformUp {
+      case q: LogicalPlan =>
+        q transformExpressionsUp {
+          case a: AttributeReference =>
+            q.resolveChildren(Seq(a.qualifier.getOrElse(""), a.name),
+              analyzer.resolver).getOrElse(a)
+        }
+    }
+
+  }
+
+  private def getIndexHints = {
+    val indexHint = QueryHint.Index.toString
+    snappySession.queryHints.collect {
+      case (hint, value) if hint.startsWith(indexHint) =>
+        val tableOrAlias = hint.substring(indexHint.length)
+        val key = catalog.lookupRelationOption(
+          catalog.newQualifiedTableName(tableOrAlias)) match {
+          case Some(relation@LogicalRelation(cf: BaseColumnFormatRelation, _, _)) =>
+            cf.table
+          case _ => tableOrAlias
+        }
+
+        val index = catalog.lookupRelation(
+          snappySession.getIndexTable(catalog.newQualifiedTableName(value)))
+
+        (key, index)
+    }
+  }
+
+}
+
+/**
+  * Replace table with index if colocation criteria is satisfied.
+  */
+case class ResolveIndex(snappySession: SnappySession) extends Rule[LogicalPlan]
+    with PredicateHelper {
+
+  lazy val catalog = snappySession.sessionState.catalog
+
+  lazy val analyzer = snappySession.sessionState.analyzer
+
+  override def apply(plan: LogicalPlan): LogicalPlan = replaceWithIndex(plan, { optimizedPlan =>
+
+    optimizedPlan.flatMap {
+      case ExtractEquiJoinKeys(joinType, leftKeys, rightKeys, extraConditions, left, right) =>
+        (left, right) match {
+          case HasColocatedEntities(colocatedEntities, leftReplacements, rightReplacements) =>
+
+            val joinKeys = leftKeys.zip(rightKeys).flatMap { case (leftExp, rightExp) =>
```

Comment:
I got mislead by the name partitioningColumn in ColumnFormatRelation populated during create table.
On second thought, I suppose we need to do a more exhaustive check of expression equivalence between gemxd style expr and spark style "catalyst.expressions.Expression". This code for now should expect AttributedReference and do a strict column name check. Will alter it accordingly.